Source/Tools/FEXServer/Unwind/ELFMapping.cpp

```diff
@@ -0,0 +1,429 @@
+        DefinedSymbol->Name = Name;
+        DefinedSymbol->SectionIndex = Symbol->st_shndx;
+        Mapping->SymbolMap[DefinedSymbol->Name] = DefinedSymbol;
+        Mapping->SymbolMapByAddress[DefinedSymbol->Address] = DefinedSymbol;
+      }
+    }
+
+    Elf32_Shdr const *StrHeader = Mapping->SectionHeaders.at(Mapping->Header._32.e_shstrndx)._32;
+    char const *SHStrings = &RawBase[StrHeader->sh_offset];
+    for (uint32_t i = 0; i < Mapping->SectionHeaders.size(); ++i) {
+      Elf32_Shdr const *hdr = Mapping->SectionHeaders.at(i)._32;
+      if (strcmp(&SHStrings[hdr->sh_name], ".eh_frame_hdr") == 0) {
+        auto eh_frame_hdr = &RawBase[hdr->sh_offset];
+        // we only handle this specific unwind table encoding
+        if (eh_frame_hdr[0] == 1 && eh_frame_hdr[1] == 0x1B && eh_frame_hdr[2] == 0x3 && eh_frame_hdr[3] == 0x3b) {
+          // ptr enc : 4 bytes, signed, pcrel
+          // fde count : 4 bytes udata
+          // table enc : 4 bytes, signed, datarel
+          int fde_count = *(int*)(eh_frame_hdr + 8);
+          Mapping->UnwindEntries.clear();
+          Mapping->UnwindEntries.reserve(fde_count);
+
+          struct entry {
+            int32_t pc;
+            int32_t fde;
+          };
+
+          entry *Table = (entry*)(eh_frame_hdr+12);
+          for (int f = 0; f < fde_count; f++) {
+            uintptr_t Entry = (uintptr_t)(Table[f].pc + hdr->sh_offset);
+            Mapping->UnwindEntries.push_back(Entry);
+          }
+        }
+        break;
+      }
+    }
+
+    // Invalid mapping for top end search.
+    Mapping->SymbolMapByAddress[~0ULL] = {};
+  }
+
+  void ParseELFSymbols(ELFMemMapping *Mapping) {
+    // Find the ELF header, which should be the first element.
+    uint64_t Bits = IsELFBits(Mapping);
+    if (Bits != 0) {
+      if (Bits == 64) {
+        ParseELFSymbols64(Mapping);
+      }
+      if (Bits == 32) {
+        ParseELFSymbols32(Mapping);
+      }
+
+      return;
+    }
+  }
+
+  ELFMemMapping *LoadELFMapping(const FileMapping::FileMapping *Mapping, int FD) {
+    if (FD == -1) {
+      return nullptr;
+    }
+
+    ELFMemMapping *RemapMapping = new ELFMemMapping{};
+
+    RemapMapping->FD = FD;
+
+    struct stat buf{};
+    if (fstat(FD, &buf) != 0) {
+        return nullptr;
+    }
+
+    RemapMapping->OriginalBase = Mapping->Begin;
+    RemapMapping->TotalSize = buf.st_size;
+
+    RemapMapping->BaseRemapped = reinterpret_cast<char*>(mmap(nullptr, RemapMapping->TotalSize, PROT_READ, MAP_SHARED, FD, 0));
+
+    ParseELFSymbols(RemapMapping);
+
+    return RemapMapping;
+  }
+
+  ELFSymbol *GetSymbolFromAddress(ELFMemMapping *Mapping, uint64_t Addr) {
+    auto Sym = Mapping->SymbolMapByAddress.upper_bound(Addr);
+    if (Sym != Mapping->SymbolMapByAddress.begin()) {
+      --Sym;
+    }
+
+    if (Sym == Mapping->SymbolMapByAddress.end()) {
+      return nullptr;
+    }
+
+    // <= necessary for single byte sized instructions, like thunks.
```

Comment:
Is this right? In that case, `Addr` would still equal `Sym->second->Address` and hence be strictly smaller than `Sym->second->Address + Sym->second->Size`, wouldn't it?

Reply:
Weirdly yes otherwise Thunk entrypoints won't show up in the backtrace.

Reply:
Is there a coherent explanation we can find for this? Sounds like this is masking a bug elsewhere.